restapi/src/main/java/io/stargate/web/docsapi/resources/CollectionsResource.java

```diff
@@ -0,0 +1,232 @@
+        @ApiResponse(code = 401, message = "Unauthorized"),
+        @ApiResponse(code = 404, message = "Collection not found"),
+        @ApiResponse(code = 500, message = "Internal server error", response = Error.class)
+      })
+  @Path("collections/{collection-id}/upgrade")
+  @Consumes("application/json")
+  @Produces("application/json")
+  public Response upgradeCollection(
+      @ApiParam(
+              value =
+                  "The token returned from the authorization endpoint. Use this token in each request.",
+              required = true)
+          @HeaderParam("X-Cassandra-Token")
+          String token,
+      @ApiParam(value = "the namespace containing the collection to upgrade", required = true)
+          @PathParam("namespace-id")
+          String namespace,
+      @ApiParam(value = "the collection to upgrade", required = true) @PathParam("collection-id")
+          String collection,
+      @ApiParam(value = "Unwrap results", defaultValue = "false") @QueryParam("raw")
+          final boolean raw,
+      @ApiParam(
+              value = "JSON with the upgrade type",
+              required = true,
+              example = "{\"upgradeType\": \"SAI_INDEX_UPGRADE\"}")
+          String payload) {
+    return RequestHandler.handle(
+        () -> {
+          DataStore localDB = db.getDataStoreForToken(token);
+          Table table = localDB.schema().keyspace(namespace).table(collection);
+          if (table == null) {
+            return Response.status(Response.Status.NOT_FOUND)
+                .entity("Collection not found")
+                .build();
+          }
+          DocCollection request = mapper.readValue(payload, DocCollection.class);
+          DocCollection info = collectionService.getCollectionInfo(table, db);
+          if (request.getUpgradeType() == null
+              || !info.getUpgradeAvailable()
+              || info.getUpgradeType() != request.getUpgradeType()) {
+            return Response.status(Response.Status.BAD_REQUEST)
+                .entity("That collection cannot be upgraded in that manner")
+                .build();
+          }
+
+          boolean success =
+              collectionService.upgradeCollection(
+                  namespace,
+                  collection,
+                  new DocumentDB(localDB),
+                  request.getUpgradeType(),
+                  db.isDse());
+
+          if (success) {
+            table = localDB.schema().keyspace(namespace).table(collection);
+            info = collectionService.getCollectionInfo(table, db);
+
+            Object response = raw ? info : new ResponseWrapper(info);
+            return Response.status(Response.Status.OK)
+                .entity(Converters.writeResponse(response))
+                .build();
+          } else {
+            return Response.status(Response.Status.BAD_REQUEST)
+                .entity("Collection was not upgraded. Are you sure it was possible to upgrade?")
```

Comment:
> Are you sure it was possible to upgrade?

Don't we validate that beforehand?

Reply:
Yea - this error is like the "catch-all" that I really wouldn't expect to occur, but technically it could